types/react-redux/index.d.ts

```diff
@@ -9,22 +9,14 @@ import * as Redux from 'redux';
 
 type ComponentClass<P> = React.ComponentClass<P>;
 type StatelessComponent<P> = React.StatelessComponent<P>;
+type Component<P> = ComponentClass<P> | StatelessComponent<P>;
 type ReactNode = React.ReactNode;
 type Store<S> = Redux.Store<S>;
 type Dispatch<S> = Redux.Dispatch<S>;
 type ActionCreator<A> = Redux.ActionCreator<A>;
 
-interface ComponentDecorator<TOriginalProps, TOwnProps> {
-    (component: ComponentClass<TOriginalProps> | StatelessComponent<TOriginalProps>): ComponentClass<TOwnProps>;
-}
-
-/**
- * Decorator that infers the type from the original component
- *
- * Can't use the above decorator because it would default the type to {}
- */
-export interface InferableComponentDecorator {
-    <P, TComponentConstruct extends (ComponentClass<P> | StatelessComponent<P>)>(component: TComponentConstruct): TComponentConstruct;
+interface ComponentDecorator<TStateProps, TDispatchProps, TOwnProps> {
```

Comment:
In the old version, `TOriginalProps` was `TStateProps & TDispatchProps`. This meant that the `component` parameter had type `Component<TStateProps & TDispatchProps>`.

In the new version, `component` is properly `Component<TStateProps & TDispatchProps & TOwnProps>`